one-table.redshift.sql

```diff
@@ -0,0 +1,330 @@
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_boolean(input text)
+RETURNS BOOLEAN
+STABLE AS $$
+	try:
+		v = input.lower() == 'true'
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
+
+-- DROP FUNCTION _airbyte_safe_cast_to_text(input text);
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_text(input text)
+RETURNS TEXT
+STABLE AS $$
+	try:
+		v = str(input)
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
+
+-- DROP FUNCTION _airbyte_safe_cast_to_timestamp(input text);
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_timestamp(input text)
+RETURNS TIMESTAMP
+STABLE AS $$
+	import dateutil.parser
+	try:
+		v = dateutil.parser.parse(input)
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
+
+---------------------------
+--------- UTILS -----------
+---------------------------
+
+-- http://datasamurai.blogspot.com/2016/06/creating-uuid-function-in-redshift.html
+CREATE OR REPLACE FUNCTION gen_uuid()
+RETURNS VARCHAR(36)
+VOLATILE AS $$
+ import uuid
+ return uuid.uuid4().__str__()
+$$ LANGUAGE plpythonu;
+
+-------------------------------------
+--------- TYPE AND DEDUPE -----------
+-------------------------------------
+
+CREATE OR REPLACE PROCEDURE _airbyte_prepare_raw_table()
+AS $$
+BEGIN
+  CREATE SCHEMA IF NOT EXISTS x;
+  CREATE TABLE IF NOT EXISTS x.users_raw (
+      "_airbyte_raw_id" VARCHAR(36) NOT NULL, -- Airbyte column, cannot be null
+      "_airbyte_data" SUPER NOT NULL, -- Airbyte column, cannot be null
+      "_airbyte_extracted_at" timestamp NOT NULL, -- Airbyte column, cannot be null
+      "_airbyte_loaded_at" timestamp, -- Airbyte column
+      PRIMARY KEY ("_airbyte_raw_id")
+  );
+END
+$$ LANGUAGE plpgsql;
+
+
+CREATE OR REPLACE PROCEDURE _airbyte_type_dedupe()
+AS $$
+
+DECLARE missing_pk_count INTEGER DEFAULT 0;
+
+BEGIN
+
+  -- Step 1: Validate the incoming data
+  -- We can't really do this properly in the pure-SQL example here, but we should throw if any row doesn't have a PK
+  missing_pk_count := (
+    SELECT COUNT(1)
+    FROM x.USERS_RAW
+    WHERE
+      "_airbyte_loaded_at" IS NULL
+      AND _airbyte_safe_cast_to_integer(_airbyte_data.id::text) IS NULL
+    );
+
+  IF missing_pk_count > 0 THEN
+    RAISE EXCEPTION 'Table has % rows missing a primary key', missing_pk_count;
+  END IF;
+
+  -- Moving the data and deduping happens in a transaction to prevent duplicates from appearing
+  -- BEGIN
+
+  -- Step 2: Move the new data to the typed table
+  INSERT INTO x.users (
+    id,
+    first_name,
+    age,
+    updated_at,
+    address,
+    _airbyte_meta,
+    _airbyte_raw_id,
+    _airbyte_extracted_at
+  )
+
+  WITH intermediate_data AS (
+    SELECT
+       _airbyte_safe_cast_to_integer(_airbyte_data.id::text) as id
+      ,_airbyte_safe_cast_to_text(_airbyte_data.first_name::text) as first_name
+      ,_airbyte_safe_cast_to_integer(_airbyte_data.age::text) as age
+      ,_airbyte_data.address as address -- there is no way to make a JSON/SUPER safe cast in Redshift
```

Comment:
what does it mean? Address will be a JSON?

Reply:
Yep! That's what we want, and in redshift lingo "any sub-section of a SUPER-type remains a SUPER", so it's going to be valid... we just can't confirm that behavior.  Valid JSON in this case also includes strings, numbers, arrays, etc.